### ISSUE TITLE
Implement Ray's lightgbm distributed training to compare against regular lightgbm distributed

### ISSUE THREAD

**Issue by jfomhover** [microsoft/lightgbm-benchmark#173]
[Ray is a new parallelization framework: _"Ray is an open source project that makes it simple to scale any compute-intensive Python workload — from deep learning to production model serving."_.](https://docs.ray.io/en/latest/index.html)

RayML provides [a LightGBM integration](https://docs.ray.io/en/latest/lightgbm-ray.html). We'd like to compare this version against vanilla distributed LightGBM (mpi).

The expected impact of this task:
- increase the value of the benchmark for the lightgbm community
- assess Ray's viability as a library for supporting scalable ML workloads

## Learning Goals

By working on this project you'll be able to learn:
- how to use RayML
- how to compare training in a consistent benchmark

## Expected Deliverable:

To complete this task, you need to deliver:
- 1 working python script to train LightGBM on Ray

### Prepare for coding

1. Follow the [installation process](https://microsoft.github.io/lightgbm-benchmark/run/install/), please [report any issue](https://github.com/microsoft/lightgbm-benchmark/issues) you meet, that will help!
2. Clone this repo, create your own branch `username/lightgbmonray` (or something) for your own work (commit often!).
3. In `src/scripts/training` create a folder `lightgbm_on_ray/` and copy the content of [`src/scripts/samples/`](https://github.com/microsoft/lightgbm-benchmark/tree/main/src/scripts/sample) in it.

Note: it might be worth copying heavily from the lightgbm python training script.

### Local development

Let's start locally first.

To iterate on your python script, you need to consider a couple of constraints:
- Follow the instructions in the sample script to modify and make your own.
- Please consider using inputs and outputs that are provided as directories, not single files. There's a helper function to let you automatically select the unique file contained in a directory (see `src/common/io.py` function `input_file_path`)

#### Component specification

1. First, unit tests. Edit `tests/aml/test_components.py` and watch for the list of components. Add the relative path to your component spec in this list.
    You can test your component by running
    ```bash
    pytest tests/aml/test_components.py -v -k name_of_component
    ```

2. Edit the file `spec.yaml` in the directory of your component (copied from sample) and align its arguments with the expected arguments of your component until you pass the unit tests.

#### Integration in the training pipeline

WORK IN PROGRESS

**Comment by abhi-rf** [comment]
Hello,

The RayML link you provided also contains sample script for running LightGBM. How does the script to be written here should be different from the one provided ? I am willing to give the issue a try :)

**Comment by jfomhover** [comment]
> Hello,
> 
> The RayML link you provided also contains sample script for running LightGBM. How does the script to be written here should be different from the one provided ? I am willing to give the issue a try :)

Missed that comment :-/ thank you for reaching out! I'm thinking about preparing some ground work in the repo to provide the right helper class for LightGBM on Ray. @ruizhuanguw is looking forward to implement some ray tune example, and I can provide a ray sample.

**Comment by jfomhover** [comment]
Here's a first PR to implement some helper class to support Ray components in our benchmark. This also implements lightgbm inferencing using Ray. https://github.com/microsoft/lightgbm-benchmark/pull/215

**Comment by ruizhuanguw** [comment]
@jfomhover what is the advantage of using inputs and outputs that are provided as directories, instead of single files?
